espnet/nets/pytorch_backend/transformer/encoder_layer.py

```diff
@@ -73,22 +107,43 @@ def forward(self, x, mask, cache=None):
             residual = residual[:, -1:, :]
             mask = None if mask is None else mask[:, -1:, :]
 
+        if pos_emb is not None:
+            x_att = self.self_attn(x_q, x, x, pos_emb, mask)
+        else:
+            x_att = self.self_attn(x_q, x, x, mask)
+
         if self.concat_after:
-            x_concat = torch.cat((x, self.self_attn(x_q, x, x, mask)), dim=-1)
+            x_concat = torch.cat((x, x_att), dim=-1)
             x = residual + self.concat_linear(x_concat)
         else:
-            x = residual + self.dropout(self.self_attn(x_q, x, x, mask))
+            x = residual + self.dropout(x_att)
         if not self.normalize_before:
-            x = self.norm1(x)
-
+            x = self.norm_mha(x)
+
+        # convolution module
+        if self.conv_module is not None:
+            residual = x
+            if self.normalize_before:
+                x = self.norm_conv(x)
+            x = residual + self.dropout(self.conv_module(x))
+            if not self.normalize_before:
+                x = self.norm_conv(x)
+
+        # feed forward module
         residual = x
         if self.normalize_before:
-            x = self.norm2(x)
-        x = residual + self.dropout(self.feed_forward(x))
+            x = self.norm_ff(x)
+        x = residual + self.ff_scale * self.dropout(self.feed_forward(x))
         if not self.normalize_before:
-            x = self.norm2(x)
+            x = self.norm_ff(x)
+
+        if self.conv_module is not None:
+            x = self.norm_final(x)
```

Comment:
Is this necessary? (maybe redundant)

Reply:
As described In the Conformer paper Sec 2.4, they use LayrNrom not only in each module, but also at the end of the block. So I think there are five LayerNorms totally.

Reply:
I checked the paper again. Yes, it seems true. But if we use pre-norm, the first layer norm (self.norm_ff_macaron()) in the next layer corresponds to the self.norm_final() in the current layer.
Check the original implementation https://github.com/zhuohan123/macaron-net/blob/master/translation/fairseq/models/transformer.py#L595.
We need some discussions.

Reply:
Yes, I also feel a little confused about the `self.norm_final`. I think in the conformer paper, the final `LayerNorm` is different from the `self.norm_ff_macaron` in the next block, also described in eq.(1). However, in fairseq's implementation, I think they didn't use an additional `LayerNorm` for the final output. They only use `LayerNorm` before or after each module. I am not sure whether `self.norm_final` is necessary for Conformer. I can do some experiments latter.